src/replay.ts

```diff
@@ -299,6 +259,160 @@ function validateReplayNormalizedOptions(fromDate: Date, normalizers: MapperFact
+      const messageBuffer = bufferLine.slice(DATE_MESSAGE_SPLIT_INDEX + 1)
+      // as any due to https://github.com/Microsoft/TypeScript/issues/24929
+      yield {
+        localTimestamp: localTimestampBuffer,
+        message: messageBuffer
+      } as any
+      // ignore empty lines unless withDisconnects is set to true
+      // do not yield subsequent undefined messages
+    } else if (withDisconnects === true && lastMessageWasUndefined === false) {
+      lastMessageWasUndefined = true
+      yield undefined as any
+    }
+  }
+}
+
+function decodeMessage(withMicroseconds: boolean, message: { localTimestamp: Buffer; message: Buffer } | undefined): any {
+  if (!message) return message
+  const localTimestampBuffer = message.localTimestamp
+  const localTimestampString = localTimestampBuffer.toString()
+  const localTimestamp = new Date(localTimestampString)
+  if (withMicroseconds) {
+    // provide additionally fractions of millisecond at microsecond resolution
+    // local timestamp always has format like this 2019-06-01T00:03:03.1238784Z
+    localTimestamp.μs = parseμs(localTimestampString)
+  }
+
+  // reuse original message object to lower memory impact.
+  const decodedMessage: any = message
+  decodedMessage.localTimestamp = localTimestamp
+  decodedMessage.message = JSON.parse(message.message as any)
+
+  return decodedMessage
+}
+
+// we leverage the fact that localTimestamps are normalized to same format
+// can numeric digits string code points are sequential
+function compareTimestampBuffer(t1: Buffer, t2: Buffer) {
+  for (let i = 17; i < 27; i++) {
+    const delta = t1[i] - t2[i]
+    if (delta > 0) {
+      return 1
+    } else if (delta < 0) {
+      return -1
+    }
+  }
+  return 0
+}
+
+const MESSAGE_EMPTY = Symbol('MESSAGE_EMPTY')
+const MESSAGE_DONE = Symbol('MESSAGE_DONE')
+const enum Slot {
+  ITERATOR,
+  VALUE,
+  TIMESTAMP
+}
+type MessageSlot = [AsyncGenerator, any, Buffer | undefined]
+async function* mergeMessageStreams(streams: AsyncGenerator[]) {
+  let slots: MessageSlot[] = streams.map((iterator) => [iterator, MESSAGE_EMPTY, undefined])
+  let len = slots.length
+  let needCleanup = false
+
+  function fillSlots() {
+    const promises = slots.map((slot) => {
+      const iterator = slot[Slot.ITERATOR]
+      if (slot[Slot.VALUE] === MESSAGE_EMPTY) {
+        return iterator.next().then((result) => {
+          if (result.done) {
+            slot[Slot.VALUE] = MESSAGE_DONE
+            needCleanup = true
+          } else {
+            slot[Slot.VALUE] = result.value
+            slot[Slot.TIMESTAMP] = slot[Slot.VALUE] ? slot[Slot.VALUE].localTimestamp : undefined
+          }
+        })
+      } else {
+        return
+      }
+    })
+    return Promise.all(promises)
+  }
+
+  function cleanupDoneSlots() {
+    if (needCleanup) {
+      slots = slots.filter((slot) => slot[Slot.VALUE] !== MESSAGE_DONE)
+      len = slots.length
+      needCleanup = false
+    }
+  }
+
```

Comment:
I have some idea to implement `pickSlot` in more performant way.

Idea is to leverage the fact that there's higher probability that if one stream is emitting, it'll continue to emit sequentially without the need to switch source stream. So we can retain a ref to previous slot between yield, check against it first to see if it's next localTimestamp is still better than recorded second-earliest localTimestamp. If true, then we can short-circuit return, don't need to check across all slots.

But this functional `.reduce` way is more easy to understand. It looks nicer.